mlir-clang/Lib/clang-mlir.cc

```diff
@@ -38,6 +38,67 @@ using namespace mlir;
 
 #define DEBUG_TYPE "clang-mlir"
 
+static cl::opt<bool>
+    memRefFullRank("memref-fullrank", cl::init(false),
+                   cl::desc("Get the full rank of the memref."));
+
+/// Try to typecast the caller arg of type MemRef to fit the corresponding
+/// callee arg type. We only deal with the cast where src and dst have the same
+/// shape size and elem type, and just the first shape differs: src has -1 and
+/// dst has a constant integer.
+static mlir::Value castCallerMemRefArg(mlir::Value callerArg,
+                                       mlir::Type calleeArgType,
+                                       mlir::OpBuilder &b) {
+  mlir::OpBuilder::InsertionGuard guard(b);
+  mlir::Type callerArgType = callerArg.getType();
+
+  if (MemRefType dstTy = calleeArgType.dyn_cast_or_null<MemRefType>()) {
+    MemRefType srcTy = callerArgType.dyn_cast<MemRefType>();
+    if (srcTy && dstTy.getElementType() == srcTy.getElementType()) {
+      auto srcShape = srcTy.getShape();
+      auto dstShape = dstTy.getShape();
+
+      if (srcShape.size() == dstShape.size() && !srcShape.empty() &&
+          srcShape[0] == -1 &&
```

Comment:
I assume here you are looking if the memref has static shape, right? If so I would use `hasStaticShape` (see BuiltinTypes.h) just to be more explicit (-1 looks confusing maybe).

Reply:
Thanks Lorenzo, perhaps I'm trying to do another thing here - I'm checking if the `srcTy` is like `memref<?xNxMxf32>` while the `dstTy` is `memref<KxNxMxf32>`, that is, `dstTy` is almost like `srcTy`, while just in the first dimension `srcTy` is undetermined. Does this make sense?